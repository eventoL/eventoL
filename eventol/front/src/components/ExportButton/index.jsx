import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import {CSVLink} from 'react-csv';

import './index.scss';

export default class ExportButton extends React.PureComponent {
  static propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape()),
    filename: PropTypes.string,
    label: PropTypes.string,
    type: PropTypes.oneOf([
      'default',
      'primary',
      'success',
      'info',
      'warning',
      'danger',
      'link',
    ]),
  };

  static defaultProps = {
    data: [],
    filename: 'export',
    label: '',
    type: 'default',
  };

  state = {
    header: [],
    rows: [],
    totals: [],
  };

  getField = (columns, field) => {
    const values = [];
    columns.forEach(column => {
      if (!column.hasOwnProperty('Header')) return null;
      if (!column.hasOwnProperty('columns')) values.push(column[field]);
      else column.columns.forEach(subcolumn => values.push(subcolumn[field]));
      return column;
    });
    return values;
  };

  getRows(columns) {
    const {data} = this.props;
    const accessors = this.getField(columns, 'accessor');
    return data.map(row =>
      accessors.map(accessor => this.runAccessor(accessor, row))
    );
  }

  runAccessor = (accessor, row) => {
    if (typeof accessor === 'string') return row[accessor];
    return accessor(row);
  };

  updateCsv(columns) {
    const {data} = this.props;
    if (!_.isEmpty(data)) {
      const header = this.getField(columns, 'Header');
      const rows = this.getRows(columns);
      const totals = this.getField(columns, 'total');
      this.setState({header, rows, totals});
    }
  }

  render() {
    const {type, label, filename} = this.props;
    const {header, rows, totals} = this.state;
    const csvData = [header, ...rows, totals];
    return (
      <div className={`export btn btn-raised btn-${type}`}>
        <CSVLink data={csvData} filename={`${filename}.csv`}>
          {label}
        </CSVLink>
      </div>
    );
  }
}
