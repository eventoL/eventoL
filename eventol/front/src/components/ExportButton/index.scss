@import '../../styles/variables';

.export {
  a {
    color: $a-color;
    text-decoration: none;
  }
}
