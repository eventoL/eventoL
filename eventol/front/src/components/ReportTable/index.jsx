import React from 'react';
import PropTypes from 'prop-types';
import ReactTable from 'react-table';

import 'react-table/react-table.css';


export default class ReportTable extends React.Component {
  static propTypes = {
    count: PropTypes.number,
    data: PropTypes.array,
    defaultRows: PropTypes.number,
    eventsPrivateData: PropTypes.object,
    exportButton: PropTypes.object,
    fetchData: PropTypes.func,
    loading: PropTypes.bool,
    pages: PropTypes.number,
    table: PropTypes.string,
    totals: PropTypes.object,
  };

  getEventColumns(){
    const {count} = this.props;
    return {
      Header: gettext('Event'),
      columns: [{
        Header: gettext('Name'), accessor: 'name', resizable: false, sortable: true, total: `${gettext('Events')}: ${count}`,
        minWidth: 150, Footer: (<span><strong>{gettext('Events')}:  </strong>{count}</span>)
      }]
    };
  }

  getOrganizersColumns(){
    return {
      Header: gettext('Organizers'),
      columns: [
        {Header: gettext('Email'), accessor: 'email'},
        {Header: gettext('Organizers'), accessor: 'organizers'}
      ]};
  }

  getLocationColumns(){
    return {
      Header: gettext('Location'),
      columns: [{
        Header: gettext('Address detail'), id: 'address_detail',
        accessor: ({locationDetail: {address_detail}}) => address_detail
      },{
        Header: gettext('Address'), id: 'address',
        accessor: ({locationDetail: {address}}) => address
      }, {
        Header: gettext('Province'), id: 'province',
        accessor: ({locationDetail: {province}}) => province
      }]
    };
  }

  getAssitanceConfirmatedColumns(){
    const {totals} = this.props;
    if (!totals.hasOwnProperty('attendees')) return {};
    return {
      Header: gettext('Assistance (confirmed)'),
      columns: [{
        Header: gettext('Attendees'), id: 'attendees',
        accessor: ({assistanceDetail: {attendees}}) => attendees.confirmed, total: totals.attendees.confirmed,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.attendees.confirmed}</span>)
      },{
        Header: gettext('Organizers'), id: 'organizer',
        accessor: ({report: {organizer}}) => organizer.confirmed, total: totals.organizers.confirmed,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.organizers.confirmed}</span>)
      },{
        Header: gettext('Collaborators'), id: 'collaborator',
        accessor: ({report: {collaborator}}) => collaborator.confirmed, total: totals.collaborators.confirmed,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.collaborators.confirmed}</span>)
      },{
        Header: gettext('Installers'), id: 'installer',
        accessor: ({report: {installer}}) => installer.confirmed, total: totals.installers.confirmed,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.installers.confirmed}</span>)
      },{
        Header: gettext('Speakers'),id: 'speakers',
        accessor: ({report: {speakers}}) => speakers, total: totals.speakers,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.speakers}</span>)
      }]
    };
  }

  getActivitiesColumns(){
    const {totals} = this.props;
    if (!totals.hasOwnProperty('activities')) return {};
    return {
      Header: gettext('Activities'),
      columns: [{
        Header: gettext('Activities'), id: 'activities',
        accessor: ({report: {activity: {total}}}) => total, total: totals.activities.confirmed,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.activities.confirmed}</span>)
      },{
        Header: gettext('Installations'), id: 'installations',
        accessor: ({report: {installation: {total}}}) => total, total: totals.installations.total,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.installations.total}</span>)
      }]
    };
  }

  getConfirmationColumnsBy(category, total_category, field, title){
    const {totals} = this.props;
    if (!totals.hasOwnProperty(total_category)) return {};
    return {
      Header: title,
      columns: [{
        Header: gettext('Confirmed'), id: `${title.toLowerCase()}_confirmed`,
        accessor: data => data[field][category].confirmed, total: totals[total_category].confirmed,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals[total_category].confirmed}</span>)
      },{
        Header: gettext('Not Confirmed'), id: `${title.toLowerCase()}_not_confirmed`,
        accessor: data => data[field][category].not_confirmed, total: totals[total_category].not_confirmed,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals[total_category].not_confirmed}</span>)
      },{
        Header: gettext('Total'), id: `${title.toLowerCase()}_total`,
        accessor: data => data[field][category].total, total: totals[total_category].total,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals[total_category].total}</span>)
      }]
    };
  }

  getAssistancesFullColumns(){
    const {totals} = this.props;
    if (!totals.hasOwnProperty('attendees')) return {};
    return {
      attendees: this.getConfirmationColumnsBy('attendees', 'attendees', 'assistanceDetail', 'Attendees'),
      collaborators: this.getConfirmationColumnsBy('collaborator', 'collaborators', 'report', 'Collaborators'),
      installers: this.getConfirmationColumnsBy('installer', 'installers', 'report', 'Installers'),
      organizers: this.getConfirmationColumnsBy('organizer', 'organizers', 'report', 'Organizers'),
      speakers: {
        Header: gettext('Speakers'),id: 'speakers',
        accessor: ({report: {speakers}}) => speakers, total: totals.speakers,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.speakers}</span>)
      }
    };
  }

  getSoftwaresColumns(){
    const {totals} = this.props;
    if (!totals.hasOwnProperty('installations')) return [];
    return Object.keys(totals.installations.softwares).map(software => {
      return {
        Header: software, id: `installations_${software}`,
        accessor: ({report: {installation: {software_count}}}) => software_count[software] || 0, total: totals.installations.softwares[software] || 0,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.installations.softwares[software] || 0}</span>)
      };
    });
  }

  getInstallationsColumns(){
    const {totals} = this.props;
    if (!totals.hasOwnProperty('installations')) return {};
    const softwaresColumns = this.getSoftwaresColumns();
    return {
      Header: gettext('Installations'),
      columns: [{
        Header: gettext('Quantity'), id: 'installations_quantity',
        accessor: ({report: {installation: {total}}}) => total, total: totals.installations.total,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.installations.total}</span>)
      },
      ...softwaresColumns]
    };
  }

  getActivitiesFieldsColumns(category, plural_category, names){
    const {totals} = this.props;
    if (!totals.hasOwnProperty('activities')) return [];
    return Object.keys(totals.activities.details[plural_category]).map(element => {
      const Header = (element !== 'None') ? names[element - 1] : gettext('Don\'t configured');
      return {
        Header, id: `activity_${category}_${element}`,
        accessor: ({report: {activity}}) => activity[`${category}_count`][element] || 0, total: totals.activities.details[plural_category][element] || 0,
        Footer: (<span><strong>{gettext('Total')}: </strong>{totals.activities.details[plural_category][element] || 0}</span>)
      };
    });
  }

  getActivitiesFullColumns(){
    const confirmationColumn = this.getConfirmationColumnsBy('activity', 'activities', 'report', 'Attendees');
    if (!confirmationColumn.hasOwnProperty('columns')) return [];
    const statusColumns = this.getActivitiesFieldsColumns('status', 'status', [gettext('Proposal'), gettext('Accepted'), gettext('Rejected')]);
    const typeColumns = this.getActivitiesFieldsColumns('type', 'types', [gettext('Talk'), gettext('Workshop'), gettext('Lightning talk'), gettext('Other')]);
    const levelColumns = this.getActivitiesFieldsColumns('level', 'levels', [gettext('Beginner'), gettext('Medium'), gettext('Advanced')]);
    confirmationColumn.columns = [
      ...confirmationColumn.columns,
      ...statusColumns,
      ...typeColumns,
      ...levelColumns,
    ];
    return confirmationColumn;
  }

  getColumns(){
    const {table, eventsPrivateData} = this.props;
    const eventColumns = this.getEventColumns();
    const locationColumns = this.getLocationColumns();
    const assistanceConfirmatedColumns = this.getAssitanceConfirmatedColumns();
    const activitiesColumns = this.getActivitiesColumns();
    const assistancesFullColumns = this.getAssistancesFullColumns();
    const installationsColumns = this.getInstallationsColumns();
    const activitiesFullColumns = this.getActivitiesFullColumns();
    const columns = [eventColumns, locationColumns];
    if (eventsPrivateData){
      const organizersColumns = this.getOrganizersColumns();
      columns.push(organizersColumns);
    }
    if (table === 'confirmed'){
      columns.push(assistanceConfirmatedColumns);
      columns.push(activitiesColumns);
    }
    if (table === 'installations'){
      columns.push(assistancesFullColumns.installers);
      columns.push(installationsColumns);
    }
    if (table === 'assitance'){
      columns.push(assistancesFullColumns.attendees);
      columns.push(assistancesFullColumns.collaborators);
      columns.push(assistancesFullColumns.installers);
      columns.push(assistancesFullColumns.organizers);
      columns.push(assistancesFullColumns.speakers);
    }
    if (table === 'activities'){
      columns.push(assistancesFullColumns.speakers);
      columns.push(activitiesFullColumns);
    }
    return columns;
  }

  render(){
    const {
      data, pages, loading, defaultRows, fetchData, exportButton,
    } = this.props;
    const columns = this.getColumns();
    if (exportButton) exportButton.updateCsv(columns);
    return (
      <ReactTable
        className='-striped -highlight'
        columns={columns}
        data={data}
        defaultPageSize={defaultRows}
        defaultSorted={[{id: 'name', desc: false}]}
        loading={loading}
        manual
        multiSort={false}
        noDataText={gettext('There isn\'t any event yet.')}
        onFetchData={fetchData}
        pageSizeOptions={[5, 10, 15, 20, 25, 50, 100]}
        pages={pages}
        sortable={false}
      />
    );
  }
}
