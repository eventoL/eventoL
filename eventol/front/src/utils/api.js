import _ from 'lodash';
import Cookies from 'js-cookie';

import Logger from './logger';
import {getReportUrl} from './urls';

export const getCsrf = () => Cookies.get('csrftoken');

export const addQueryString = (url, query) => {
  const queryKeys = query ? Object.keys(query) : [];
  if (!_.isEmpty(queryKeys)) {
    const params = queryKeys.map(k => `${k}=${encodeURIComponent(query[k])}`);
    return `${url}?${params.join('&')}`;
  }
  return url;
};

export const genericFetch = (url, queryString, params) => {
  const newUrl = addQueryString(url, queryString);
  return fetch(newUrl, params)
    .then(res => {
      if (res.ok) {
        return res.json();
      }
      return Promise.reject(res);
    })
    .catch(res => {
      if (res.status === 403) {
        window.location.hash = '/';
        window.location.href = '/';
        Logger.warning('Redirect to /', res.status, res.statusText);
      } else {
        Logger.error('There has been an error', res.status, res.statusText);
      }
    });
};

export const getUrl = (url, queryString) =>
  genericFetch(url, queryString, {
    method: 'GET',
    credentials: 'same-origin',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      'X-CSRFToken': getCsrf(),
    },
  });

export const postUrl = (url, data, queryString) =>
  genericFetch(url, queryString, {
    method: 'POST',
    headers: {
      Accept: 'application/json, text/plain, */*',
      'Content-Type': 'application/json',
      'X-CSRFToken': getCsrf(),
    },
    body: JSON.stringify(data),
    credentials: 'same-origin',
  });

export const loadReports = (pageSize, page, sorted) =>
  getUrl(getReportUrl(pageSize, page, sorted));
