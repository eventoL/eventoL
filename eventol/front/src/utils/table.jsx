import React from 'react';

export const getOrganizersColumns = () => ({
  Header: gettext('Organizers'),
  columns: [
    {Header: gettext('Email'), accessor: 'email'},
    {Header: gettext('Organizers'), accessor: 'organizers'},
  ],
});

export const getLocationColumns = () => ({
  Header: gettext('Location'),
  columns: [
    {
      Header: gettext('Address detail'),
      id: 'address_detail',
      accessor: ({locationDetail: {address_detail: addressDetail}}) =>
        addressDetail,
    },
    {
      Header: gettext('Address'),
      id: 'address',
      accessor: ({locationDetail: {address}}) => address,
    },
    {
      Header: gettext('Province'),
      id: 'province',
      accessor: ({locationDetail: {province}}) => province,
    },
  ],
});

export const getEventColumns = count => ({
  Header: gettext('Event'),
  columns: [
    {
      Header: gettext('Name'),
      accessor: 'name',
      resizable: false,
      sortable: true,
      total: `${gettext('Events')}: ${count}`,
      minWidth: 150,
      Footer: (
        <span>
          <strong>{`${gettext('Events')}: `}</strong>
          {count}
        </span>
      ),
    },
  ],
});

export const getAssitanceConfirmatedColumns = totals => {
  if (!totals.hasOwnProperty('attendees')) return {};
  return {
    Header: gettext('Assistance (confirmed)'),
    columns: [
      {
        Header: gettext('Attendees'),
        id: 'attendees',
        accessor: ({assistanceDetail: {attendees}}) => attendees.confirmed,
        total: totals.attendees.confirmed,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals.attendees.confirmed}
          </span>
        ),
      },
      {
        Header: gettext('Organizers'),
        id: 'organizer',
        accessor: ({report: {organizer}}) => organizer.confirmed,
        total: totals.organizers.confirmed,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals.organizers.confirmed}
          </span>
        ),
      },
      {
        Header: gettext('Collaborators'),
        id: 'collaborator',
        accessor: ({report: {collaborator}}) => collaborator.confirmed,
        total: totals.collaborators.confirmed,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals.collaborators.confirmed}
          </span>
        ),
      },
      {
        Header: gettext('Installers'),
        id: 'installer',
        accessor: ({report: {installer}}) => installer.confirmed,
        total: totals.installers.confirmed,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals.installers.confirmed}
          </span>
        ),
      },
      {
        Header: gettext('Speakers'),
        id: 'speakers',
        accessor: ({report: {speakers}}) => speakers,
        total: totals.speakers,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals.speakers}
          </span>
        ),
      },
    ],
  };
};

export const getActivitiesColumns = totals => {
  if (!totals.hasOwnProperty('activities')) return {};
  return {
    Header: gettext('Activities'),
    columns: [
      {
        Header: gettext('Activities'),
        id: 'activities',
        accessor: ({
          report: {
            activity: {total},
          },
        }) => total,
        total: totals.activities.confirmed,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals.activities.confirmed}
          </span>
        ),
      },
      {
        Header: gettext('Installations'),
        id: 'installations',
        accessor: ({
          report: {
            installation: {total},
          },
        }) => total,
        total: totals.installations.total,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals.installations.total}
          </span>
        ),
      },
    ],
  };
};

export const getConfirmationColumnsBy = (
  totals,
  category,
  totalCategory,
  field,
  title
) => {
  if (!totals.hasOwnProperty(totalCategory)) return {};
  return {
    Header: title,
    columns: [
      {
        Header: gettext('Confirmed'),
        id: `${title.toLowerCase()}_confirmed`,
        accessor: data => data[field][category].confirmed,
        total: totals[totalCategory].confirmed,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals[totalCategory].confirmed}
          </span>
        ),
      },
      {
        Header: gettext('Not Confirmed'),
        id: `${title.toLowerCase()}_not_confirmed`,
        accessor: data => data[field][category].not_confirmed,
        total: totals[totalCategory].not_confirmed,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals[totalCategory].not_confirmed}
          </span>
        ),
      },
      {
        Header: gettext('Total'),
        id: `${title.toLowerCase()}_total`,
        accessor: data => data[field][category].total,
        total: totals[totalCategory].total,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals[totalCategory].total}
          </span>
        ),
      },
    ],
  };
};

export const getAssistancesFullColumns = totals => {
  if (!totals.hasOwnProperty('attendees')) return {};
  return {
    attendees: getConfirmationColumnsBy(
      totals,
      'attendees',
      'attendees',
      'assistanceDetail',
      'Attendees'
    ),
    collaborators: getConfirmationColumnsBy(
      totals,
      'collaborator',
      'collaborators',
      'report',
      'Collaborators'
    ),
    installers: getConfirmationColumnsBy(
      totals,
      'installer',
      'installers',
      'report',
      'Installers'
    ),
    organizers: getConfirmationColumnsBy(
      totals,
      'organizer',
      'organizers',
      'report',
      'Organizers'
    ),
    speakers: {
      Header: gettext('Speakers'),
      id: 'speakers',
      accessor: ({report: {speakers}}) => speakers,
      total: totals.speakers,
      Footer: (
        <span>
          <strong>{`${gettext('Total')}: `}</strong>
          {totals.speakers}
        </span>
      ),
    },
  };
};

export const getSoftwaresColumns = totals => {
  if (!totals.hasOwnProperty('installations')) return [];
  return Object.keys(totals.installations.softwares).map(software => {
    return {
      Header: software,
      id: `installations_${software}`,
      accessor: ({
        report: {
          installation: {software_count: softwareCount},
        },
      }) => softwareCount[software] || 0,
      total: totals.installations.softwares[software] || 0,
      Footer: (
        <span>
          <strong>{`${gettext('Total')}: `}</strong>
          {totals.installations.softwares[software] || 0}
        </span>
      ),
    };
  });
};

export const getInstallationsColumns = totals => {
  if (!totals.hasOwnProperty('installations')) return {};
  const softwaresColumns = getSoftwaresColumns(totals);
  return {
    Header: gettext('Installations'),
    columns: [
      {
        Header: gettext('Quantity'),
        id: 'installations_quantity',
        accessor: ({
          report: {
            installation: {total},
          },
        }) => total,
        total: totals.installations.total,
        Footer: (
          <span>
            <strong>{`${gettext('Total')}: `}</strong>
            {totals.installations.total}
          </span>
        ),
      },
      ...softwaresColumns,
    ],
  };
};

export const getActivitiesFieldsColumns = (
  totals,
  category,
  pluralCategory,
  names
) => {
  if (!totals.hasOwnProperty('activities')) return [];
  return Object.keys(totals.activities.details[pluralCategory]).map(element => {
    const Header =
      element !== 'None' ? names[element - 1] : gettext("Don't configured");
    return {
      Header,
      id: `activity_${category}_${element}`,
      accessor: ({report: {activity}}) =>
        activity[`${category}_count`][element] || 0,
      total: totals.activities.details[pluralCategory][element] || 0,
      Footer: (
        <span>
          <strong>{`${gettext('Total')}: `}</strong>
          {totals.activities.details[pluralCategory][element] || 0}
        </span>
      ),
    };
  });
};

export const getActivitiesFullColumns = totals => {
  const confirmationColumn = getConfirmationColumnsBy(
    totals,
    'activity',
    'activities',
    'report',
    'Attendees'
  );
  if (!confirmationColumn.hasOwnProperty('columns')) return [];
  const statusColumns = getActivitiesFieldsColumns(totals, 'status', 'status', [
    gettext('Proposal'),
    gettext('Accepted'),
    gettext('Rejected'),
  ]);
  const typeColumns = getActivitiesFieldsColumns(totals, 'type', 'types', [
    gettext('Talk'),
    gettext('Workshop'),
    gettext('Lightning talk'),
    gettext('Other'),
  ]);
  const levelColumns = getActivitiesFieldsColumns(totals, 'level', 'levels', [
    gettext('Beginner'),
    gettext('Medium'),
    gettext('Advanced'),
  ]);
  confirmationColumn.columns = [
    ...confirmationColumn.columns,
    ...statusColumns,
    ...typeColumns,
    ...levelColumns,
  ];
  return confirmationColumn;
};

export const getColumns = (table, eventsPrivateData, count, totals) => {
  const eventColumns = getEventColumns(count);
  const locationColumns = getLocationColumns();
  const assistanceConfirmatedColumns = getAssitanceConfirmatedColumns(totals);
  const activitiesColumns = getActivitiesColumns(totals);
  const assistancesFullColumns = getAssistancesFullColumns(totals);
  const installationsColumns = getInstallationsColumns(totals);
  const activitiesFullColumns = getActivitiesFullColumns(totals);
  const columns = [eventColumns, locationColumns];
  if (eventsPrivateData) {
    const organizersColumns = getOrganizersColumns();
    columns.push(organizersColumns);
  }
  if (table === 'confirmed') {
    columns.push(assistanceConfirmatedColumns);
    columns.push(activitiesColumns);
  }
  if (table === 'installations') {
    columns.push(assistancesFullColumns.installers);
    columns.push(installationsColumns);
  }
  if (table === 'assitance') {
    columns.push(assistancesFullColumns.attendees);
    columns.push(assistancesFullColumns.collaborators);
    columns.push(assistancesFullColumns.installers);
    columns.push(assistancesFullColumns.organizers);
    columns.push(assistancesFullColumns.speakers);
  }
  if (table === 'activities') {
    columns.push(assistancesFullColumns.speakers);
    columns.push(activitiesFullColumns);
  }
  return columns;
};
