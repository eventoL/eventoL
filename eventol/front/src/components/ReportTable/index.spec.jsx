import wait from 'waait';
import React from 'react';
import renderer from 'react-test-renderer';

jest.mock('react-table', () => 'ReactTable');
jest.mock('../ExportButton', () =>
  jest.fn(() => ({
    updateCsv: jest.fn(),
  }))
);
jest.mock('../../utils/table', () => ({
  getColumns: jest.fn(),
}));

import ReportTable from '.';
import ExportButton from '../ExportButton';
import {getColumns} from '../../utils/table';
import {
  eventsData,
  eventsPrivateData,
  expectedTotals,
} from '../../utils/__mock__/report';

describe('Report', () => {
  let tree;
  let table;
  let count;
  let columns;
  let element;
  let exportButton;
  let component;
  let communicator;

  beforeEach(() => {
    count = 0;
    table = 'confirmed';
    columns = 'columns';
    exportButton = new ExportButton();
    getColumns.mockImplementation(() => columns);

    element = (
      <ReportTable
        communicator={communicator}
        count={0}
        data={eventsData}
        defaultRows={15}
        eventsPrivateData={eventsPrivateData}
        exportButton={exportButton}
        fetchData={jest.fn()}
        isLoading
        pages={null}
        table={table}
        totals={expectedTotals}
      />
    );
    component = renderer.create(element);
    tree = component.toJSON();
  });

  describe('Snapshots', () => {
    test('Default', async () => {
      expect(tree).toMatchSnapshot();

      component.update(element);
      await wait(0);
      tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should calls getColumns function with correct params', () => {
      expect(getColumns).toBeCalled();
      expect(getColumns).toBeCalledWith(
        table,
        eventsPrivateData,
        count,
        expectedTotals
      );
    });

    test('should calls exportButton.updateCsv function with correct params', () => {
      expect(exportButton.updateCsv).toBeCalled();
      expect(exportButton.updateCsv).toBeCalledWith(columns);
    });
  });
});
