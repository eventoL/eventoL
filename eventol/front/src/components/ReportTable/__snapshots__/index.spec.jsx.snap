// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Snapshots Default 1`] = `
<ReactTable
  className="-striped -highlight"
  columns="columns"
  data={
    Array [
      Object {
        "event_slug": "event_slug",
        "id": 1,
        "location": Array [
          "Capital Federal",
          "Pompeya",
          "Buenos Aires",
          "Argentina",
        ],
        "name": "event name",
        "report": Object {
          "activity": Object {
            "confirmed": 1,
            "level_count": Object {
              "1": 3,
              "2": 2,
              "3": 1,
            },
            "not_confirmed": 2,
            "status_count": Object {
              "1": 2,
              "2": 1,
            },
            "total": 3,
            "type_count": Object {
              "Talk": 4,
              "Workshop": 2,
            },
          },
          "attendee": Object {
            "with_event_user": Object {
              "confirmed": 1,
              "not_confirmed": 2,
              "total": 3,
            },
            "without_event_user": Object {
              "confirmed": 2,
              "not_confirmed": 3,
              "total": 5,
            },
          },
          "collaborator": Object {
            "confirmed": 3,
            "not_confirmed": 4,
            "total": 7,
          },
          "event_user": Object {
            "confirmed": 4,
            "not_confirmed": 5,
            "total": 9,
          },
          "installation": Object {
            "hardware_count": Object {
              "notebook": 1,
              "phone": 2,
            },
            "installer_count": Object {
              "1": 2,
              "2": 1,
            },
            "software_count": Object {
              "arch": 1,
              "fedora": 2,
              "ubuntu": 3,
            },
            "total": 2,
          },
          "installer": Object {
            "confirmed": 6,
            "not_confirmed": 7,
            "total": 13,
          },
          "organizer": Object {
            "confirmed": 7,
            "not_confirmed": 8,
            "total": 15,
          },
          "speakers": 9,
        },
        "tags": Object {
          "name": "tag name",
          "slug": "tag slug",
        },
      },
      Object {
        "event_slug": "event_2_slug",
        "id": 2,
        "location": Array [
          "Capital Federal",
          "Monserrat",
          "Buenos Aires",
          "Argentina",
        ],
        "name": "event 2 name",
        "report": Object {
          "activity": Object {
            "confirmed": 1,
            "level_count": Object {
              "1": 3,
              "2": 2,
              "3": 1,
            },
            "not_confirmed": 2,
            "status_count": Object {
              "1": 2,
              "2": 1,
            },
            "total": 3,
            "type_count": Object {
              "Talk": 4,
              "Workshop": 2,
            },
          },
          "attendee": Object {
            "with_event_user": Object {
              "confirmed": 1,
              "not_confirmed": 2,
              "total": 3,
            },
            "without_event_user": Object {
              "confirmed": 2,
              "not_confirmed": 3,
              "total": 5,
            },
          },
          "collaborator": Object {
            "confirmed": 3,
            "not_confirmed": 4,
            "total": 7,
          },
          "event_user": Object {
            "confirmed": 4,
            "not_confirmed": 5,
            "total": 9,
          },
          "installation": Object {
            "hardware_count": Object {
              "notebook": 1,
              "phone": 2,
            },
            "installer_count": Object {
              "1": 2,
              "2": 1,
            },
            "software_count": Object {
              "arch": 1,
              "fedora": 2,
              "ubuntu": 3,
            },
            "total": 2,
          },
          "installer": Object {
            "confirmed": 6,
            "not_confirmed": 7,
            "total": 13,
          },
          "organizer": Object {
            "confirmed": 7,
            "not_confirmed": 8,
            "total": 15,
          },
          "speakers": 9,
        },
        "tags": Object {
          "name": "tag 2 name",
          "slug": "tag 2 slug",
        },
      },
    ]
  }
  defaultPageSize={15}
  defaultSorted={
    Array [
      Object {
        "desc": false,
        "id": "name",
      },
    ]
  }
  loading={true}
  manual={true}
  multiSort={false}
  noDataText="There isn't any event yet."
  onFetchData={[MockFunction]}
  pageSizeOptions={
    Array [
      5,
      10,
      15,
      20,
      25,
      50,
      100,
    ]
  }
  pages={null}
  sortable={false}
/>
`;

exports[`Report Snapshots Default 2`] = `
<ReactTable
  className="-striped -highlight"
  columns="columns"
  data={
    Array [
      Object {
        "event_slug": "event_slug",
        "id": 1,
        "location": Array [
          "Capital Federal",
          "Pompeya",
          "Buenos Aires",
          "Argentina",
        ],
        "name": "event name",
        "report": Object {
          "activity": Object {
            "confirmed": 1,
            "level_count": Object {
              "1": 3,
              "2": 2,
              "3": 1,
            },
            "not_confirmed": 2,
            "status_count": Object {
              "1": 2,
              "2": 1,
            },
            "total": 3,
            "type_count": Object {
              "Talk": 4,
              "Workshop": 2,
            },
          },
          "attendee": Object {
            "with_event_user": Object {
              "confirmed": 1,
              "not_confirmed": 2,
              "total": 3,
            },
            "without_event_user": Object {
              "confirmed": 2,
              "not_confirmed": 3,
              "total": 5,
            },
          },
          "collaborator": Object {
            "confirmed": 3,
            "not_confirmed": 4,
            "total": 7,
          },
          "event_user": Object {
            "confirmed": 4,
            "not_confirmed": 5,
            "total": 9,
          },
          "installation": Object {
            "hardware_count": Object {
              "notebook": 1,
              "phone": 2,
            },
            "installer_count": Object {
              "1": 2,
              "2": 1,
            },
            "software_count": Object {
              "arch": 1,
              "fedora": 2,
              "ubuntu": 3,
            },
            "total": 2,
          },
          "installer": Object {
            "confirmed": 6,
            "not_confirmed": 7,
            "total": 13,
          },
          "organizer": Object {
            "confirmed": 7,
            "not_confirmed": 8,
            "total": 15,
          },
          "speakers": 9,
        },
        "tags": Object {
          "name": "tag name",
          "slug": "tag slug",
        },
      },
      Object {
        "event_slug": "event_2_slug",
        "id": 2,
        "location": Array [
          "Capital Federal",
          "Monserrat",
          "Buenos Aires",
          "Argentina",
        ],
        "name": "event 2 name",
        "report": Object {
          "activity": Object {
            "confirmed": 1,
            "level_count": Object {
              "1": 3,
              "2": 2,
              "3": 1,
            },
            "not_confirmed": 2,
            "status_count": Object {
              "1": 2,
              "2": 1,
            },
            "total": 3,
            "type_count": Object {
              "Talk": 4,
              "Workshop": 2,
            },
          },
          "attendee": Object {
            "with_event_user": Object {
              "confirmed": 1,
              "not_confirmed": 2,
              "total": 3,
            },
            "without_event_user": Object {
              "confirmed": 2,
              "not_confirmed": 3,
              "total": 5,
            },
          },
          "collaborator": Object {
            "confirmed": 3,
            "not_confirmed": 4,
            "total": 7,
          },
          "event_user": Object {
            "confirmed": 4,
            "not_confirmed": 5,
            "total": 9,
          },
          "installation": Object {
            "hardware_count": Object {
              "notebook": 1,
              "phone": 2,
            },
            "installer_count": Object {
              "1": 2,
              "2": 1,
            },
            "software_count": Object {
              "arch": 1,
              "fedora": 2,
              "ubuntu": 3,
            },
            "total": 2,
          },
          "installer": Object {
            "confirmed": 6,
            "not_confirmed": 7,
            "total": 13,
          },
          "organizer": Object {
            "confirmed": 7,
            "not_confirmed": 8,
            "total": 15,
          },
          "speakers": 9,
        },
        "tags": Object {
          "name": "tag 2 name",
          "slug": "tag 2 slug",
        },
      },
    ]
  }
  defaultPageSize={15}
  defaultSorted={
    Array [
      Object {
        "desc": false,
        "id": "name",
      },
    ]
  }
  loading={true}
  manual={true}
  multiSort={false}
  noDataText="There isn't any event yet."
  onFetchData={[MockFunction]}
  pageSizeOptions={
    Array [
      5,
      10,
      15,
      20,
      25,
      50,
      100,
    ]
  }
  pages={null}
  sortable={false}
/>
`;
